### ISSUE TITLE
General guidance by Hung-yi Lee

### ISSUE THREAD

**Issue by raychenv** [raychenv/blog#363]
https://www.youtube.com/watch?v=WeHM2xpYQpw&list=PLJV_el3uVTsMhtt7_Y6sgTHGHp1Vb2P2J&index=3
https://speech.ee.ntu.edu.tw/~hylee/ml/ml2021-course-data/overfit-v6.pdf

## Framework of ML
Training data: { x1, x2, ... , xN}
Testing data: {xN+1, xN+M}
Speech recognition
x: voice y: phoneme
Image Recognition
x: image y: soup
speaker recognition
x: vioce y: john speaker
Machine Translation
x: 痛。
y: 了解痛苦了吧

Training:
step1: function with unknown y = f theta x
step2: define loss from training data L theta
step3: optimization theta* = arg min L(theta)

Testing data: {xN+1, xN+2, ...., xN+M}
Use y = f theta x to label the testing data
{yN+1, yN+2, ..., yN+M} ----> Upload to Kaggle

## General Guide
Loss on training data
Large?                                                                                   small
model bias            optimaztion                                loss on testing data
make your                                                                large                                    small
model complex                                                        overfitting   mismatch         ideal
|                                                                               more training data/data augmentation
tradeoff----------------------------------------------make model simpler

## Model Bias
- The model is too simple
find a needle in a haystack ... but there is no needle
- Solution: redesign your model to make it more flexible
y = b + w x1 -----More feature-----> y = b + Ej56 wjxj
---> Deep Learning (More neurons, layers)
y = b + Ei*ci*sigmoid(bi + Ej wij xj)

Split your training data into training set and validation set for model selection

## Optimization issue
- Large Loss not always imply model bias There is another possiblity....
![image](https://github.com/raychenv/blog/assets/11498699/20993055-3723-470e-a53f-78b4a3e5ab3c)

## Optimization issue
- Gaining the insights from comparison
- Start from shallower networks which are easier to optimIze
- If deeper network do not obtain smaller loss on training data, then there is optimization issue
- Solution: more powerful optimization technology

## Overfitting
- Small loss on training data, large loss on testing data. Why?
An extreme example
zero training loss, but large testing loss

Solution: More training data (cannot do it in HWs)
Data augmentation (You can do that in HWs)

Constrained model
- Less parameters, sharing parameters
- Less features
- Early stopping
- Regularization

Fully-connected CNN
Constrain too much ---> back to model bias

Bias complexity trade off
![image](https://github.com/raychenv/blog/assets/11498699/3f48132d-db3f-4482-96bb-dd94ee55ca50)

Cross validation
How to split Training set?
Training set and validation set
Using the results of public testing data to select your model
You are making public set better than private set

N-fold cross validation
Training set
Training Training Val
Train Val Train
Val Train Train
------> testing set (public) testing set of private 

Mismatch
Your training and testing data have different distributions. Be aware of how data is generated
